src/java/org/apache/cassandra/db/guardrails/Guardrails.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.guardrails;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+
+import static java.lang.String.format;
+
+/**
+ * Entry point for Guardrails, storing the defined guardrails and provided a few global methods over them.
+ */
+public abstract class Guardrails
+{
+    public static final String CUSTOM_GUARDRAILS_FACTORY_PROPERTY = "cassandra.custom_guardrails_factory_class";
+
+    private static final GuardrailsConfig config = GuardrailsConfig.instance;
+
+    public static final GuardrailsFactory factory = System.getProperty(CUSTOM_GUARDRAILS_FACTORY_PROPERTY) == null
+                                                    ? new DefaultGuardrailsFactory()
+                                                    : CustomGuardrailsFactory.make(System.getProperty(CUSTOM_GUARDRAILS_FACTORY_PROPERTY));
+
```

Comment:
That can be used to replace the standard guardrails for other implementations that, for example, read the configuration for some other place, or take different actions in case the guardrail is triggered. 

The use case for what this was initially conceived is having an implementation that ignores the configuration in `cassandra.yaml` and instead reads a per-user or per-tenant configuration from some other place, and applies it using the `ClientState` to know what user is checking the guardrail.

As another example, an implementation could ignore the the configuration in `cassandra.yaml` and instead try to dynamically calculate the values for the guardrail limits based on the hardware specs, so for example it limits the number of indexes or the disk usage based on the number of available cpu cores.

Reply:
I like this pluggability feature.

Reply:
> The use case for what this was initially conceived is having an implementation that ignores the configuration in cassandra.yaml and instead reads a per-user or per-tenant configuration from some other place, and applies it using the ClientState to know what user is checking the guardrail.

The example of hierarchical quotas (reacting to `per-user or per-tenant configuration`) could always be added in the base layer.  I am also much more of a fan of adding the pluggability when needed rather than earlier and hoping we use it; I have a bias as 95% of the time I find the interfaces added for future pluggability get thrown away when the use case finally drops...

> As another example, an implementation could ignore the configuration in cassandra.yaml and instead try to dynamically calculate the values for the guardrail limits based on the hardware specs, so for example it limits the number of indexes or the disk usage based on the number of available cpu cores.

That is a config problem, why do we need to change the way we handle those configs?

Based off the examples given I am more inclined to ask we drop the classes.  The factory just deals with how we check fields, and the examples given were all about changing how things are configured... so a configuration problem and not how we check the fields problem.

> I like this pluggability feature.

To me pluggability is a hard problem and needs a lot of thought put into different types of implementations which can happen and the allowed semantics around them.  I also find that until you have around 3 drastically different implementations you don't really have a generic interface, so each time you try to on-board a new implementation you rewrite the interfaces to deal with it (Index gets updated each time we get a new index implementation in-tree).

I am a fan of pluggability, do not get me wrong, but it comes with a cost that has to be reasoned about, and right now I am not seeing the benefit for this specific interfaces; I would have different feelings if the interface was `Guardrails` as that is the interface used within the code.

Reply:
I havent dig deep into the code expect just looking at that briefly in GitHub but you tell me ... is it possible to have all the existing guardrails _plus some of my own_ ? 

In other words, if I have my own impl, do I lose all default guardrails so I have to basically re-use all of them? How is this designed? Ideally, it would be nice if my own impl could build _on top_ of the default one, instead of me cherry-picking guards I want from the default (and lets be honest here, more ofthen than not, one would like to have all of them).

Again apologies for not digging deeper into this, I just have it a shot yesterday before hitting the bed.

Reply:
@smiklosovic the purpose of having a custom guardrails factory is being able to change the behaviour of the existing guardrails, so they read their limits from some other sources or take different actions when triggered. 

Most guardrails need to evaluate info that is usually private to the place where it's triggered and not publicly exposed, such as the CQL query attributes in a `CREATE TABLE` statement or the size of a partition during compaction. So anyone wanting to add a new guardrail on a new thing should open a ticket for it, like the bunch of tickets that are in the Guardrail epic ([CASSANDRA-17146](https://issues.apache.org/jira/browse/CASSANDRA-17146)). The guardrails framework should make it relatively easy: add one or two configuration properties, add an instance of one of the types of guardrails in {{Guardrails}} using that config , and call it from the possible triggering points.

Reply:
@dcapwell it's not only about configuration, custom implementations can also define different side effects when guardrails are triggered. For example, the bit about not notifying on every activation but only time to time is a thing that could be done by custom implementations that just extend/wrap the default implementation and add that feature.

> To me pluggability is a hard problem and needs a lot of thought put into different types of implementations which can happen and the allowed semantics around them. I also find that until you have around 3 drastically different implementations you don't really have a generic interface, so each time you try to on-board a new implementation you rewrite the interfaces to deal with it (Index gets updated each time we get a new index implementation in-tree).

The point of pluggability here is allowing third parties to use their custom implementations without us needing to know the use case, and not only adding interfaces for our own usage. That's what I think was done for example with secondary indexes, and we have multiple implementations out there that have been developed out of the project, and some of them have made their way back (SASI and SAI).

> I am a fan of pluggability, do not get me wrong, but it comes with a cost that has to be reasoned about, and right now I am not seeing the benefit for this specific interfaces; I would have different feelings if the interface was Guardrails as that is the interface used within the code.

The cost of extracting interfaces for the four types of guardrail and having a factory doesn't seem too high to me. Perhaps we could make it even lower if we just provide the factory, so we get rid of `DefaultGuardrails` and custom implementors have to extend the four provided guardrails, overriding whatever they need, instead of implementing an interface? CC @jakubzytka

Reply:
> The example of hierarchical quotas (reacting to per-user or per-tenant configuration) could always be added in the base layer. I am also much more of a fan of adding the pluggability when needed rather than earlier and hoping we use it; I have a bias as 95% of the time I find the interfaces added for future pluggability get thrown away when the use case finally drops...

DataStax already has a custom implementation for its cloud offering where each tenant has its own guardrails. The proposed way of implementing basic pluggability is partially extracted from that use case, with the idea that it can be useful for that kind of extensions, among others.

Reply:
> is it possible to have all the existing guardrails plus some of my own ?

With this current logic no, the only way I see that happening is if guard rails moves event driven and offers pre/post hooks to core parts of the code. The current patch (and related patches) hard code integrations.

> The point of pluggability here is allowing third parties to use their custom implementations without us needing to know the use case

The same talking point was brought up in SSTable reader/writer (see https://www.mail-archive.com/dev@cassandra.apache.org/msg17798.html) and was requested to drop there.  If we expose something for third parties it has to be maintained and can't break between minors, this brings in a ton of problems we need to deal with in order to support that pluggability layer.

> The cost of extracting interfaces for the four types of guardrail and having a factory doesn't seem too high to me

The point I am trying to get at is the maintenance cost that comes with it.  For example we will need to change things as we add more types of guard rails, such as the track warning stuff which is a distributed version.  When we have an interface exposed to 3rd parties it limits how we can change it.  Lets say I need to add a new type of protection which doesn't fit into the existing ones, now we have to worry about what release that goes in and how not to break third parties; this is a huge cost to me.  We also need to be very clear on expected semantics, and define what is allowed, what is undefined, and what is not allowed behaviors; if I can't trust the guard rails implementation I need to use something else.

> DataStax already has a custom implementation for its cloud offering where each tenant has its own guardrails. 

Can you explain how this isn't a configuration problem?  I really don't see why the following has to be pluggable when the inputs are what we are talking about

```
class DisableFlag {
    private final String what;
    private final Predicate<ClientState> guardrailsEnabled;
    private final Predicate<ClientState> disabled;

    public void ensureEnabled(@Nullable ClientState state)
    {
    if (guardrailsEnabled.test(state) && disabled.test(state))
        fail(what + " is not allowed");
    }
}
```

Lookup of configs per state and/or thread makes sense to me, but this is just the input to the actual check logic, yet this patch makes the check logic pluggable; I am questioning why the check logic is the pluggable layer, every example given has been about configuration.

Reply:
Let me be a bit more clear.

1) if we have an API which can be changed on boot, but this API is *free to change in a minor*, I am ok with it.  Aka a hook for datastax to replace how guardrails works if they are willing to deal with minors changing the APIs; then I have no objections.
2) if the config layer can be pluggable, but we share the same check logic, I am also ok with it.  I see value with pluggable config layer

I am cool changing the APIs over time, so if there are reasons to replace the check logic, I am open to making changes; I just don't see the value currently so prefer simpler logic.

Reply:
Aren't we just discussing whether:
- That factory is exposing a public API intended for 3rd party implementations + all the costs involved in that that David mentions, like if it were a product that needs to be maintained.
vs
- That factory is just an internal implementation detail to support a handful of different implementations but there are no guarantees on compatibility, versioning, etc It's just like any other random interface in the code that can change at any time.

If it's so let's just add a disclaimer being clear on the purpose/intentions? The day we have a gazillion companies wanting to implement their guardrails version and we need to 'productize' it then we can discuss that point.

Hope it makes sense. My 2cts.

Reply:
> if the config layer can be pluggable, but we share the same check logic, I am also ok with it. I see value with pluggable config layer

[Here](https://github.com/adelapena/cassandra/commit/b69378c31869b620a2cc3aa0f8b71ec1379deec3) is a quick stab at how the pluggable config would look like. It's still incomplete and I have to better separate the MBean interface from the GuardrailConfig interface, but I guess it can serve to give us an idea of that approach. 

It allows to provide different configurations depending on the `ClientState`, which is enough for the multitenant use case and the current guardrails. However it requires that the guardrails are stateless, as they currently are. However, if we are going to add guardrails with state, such as rate-based limits, this approach wouldn't work.

> every example given has been about configuration.
Other example would be any implementation trying to change the actions that are done when a guardrail is triggered. We couldn't do that with the pluggable config, but it could be done with the pluggable guardrail factory.

Reply:
> Here is a quick stab at how the pluggable config would look like

Took a quick skim and mostly LGTM; having a mapping function to go from `state -> config` allows people to change there without really changing the semantics of how our checks actually work.

One thing I like about the patch is it returns the configs (and could be immutable); so we can have a consistent view of the configs during the life of a query (in track warnings I had to do this.. this is a very good feature); there are times where we need to check the configs multiple times within a query, so doing this with a ton of volatile configs is slow, so not having every call site figure out the cacheing... kawlz

> However it requires that the guardrails are stateless

These are accessed concurrently; they better be stateless!  Global singletons are dangerous... but this is Cassandra!

> However, if we are going to add guardrails with state, such as rate-based limits, this approach wouldn't work.

Agree, but I still question why we would want rate-limiting... Not notifying user1 because we notified user2 feels very dangerous to me (different teams often share the same credentials... just because they shouldn't doesn't mean they are not going to)...  periodically warn/fail is also super confusing to users: "wait, why did it fail the first time but not on retry?", and "why is only this one node getting a warning...".

Reply:
> * That factory is exposing a public API intended for 3rd party implementations + all the costs involved in that that David mentions, like if it were a product that needs to be maintained.
> vs
> * That factory is just an internal implementation detail to support a handful of different implementations but there are no guarantees on compatibility, versioning, etc It's just like any other random interface in the code that can change at any time.

That is my understanding.  If we are talking about public API this is a ton more work to support, if we are talking about an implementation detail then we should focus on simplify and remove complications; I can bend a little on "public API" assuming we offer 0 support cross major/minor/hotfix releases.

> The day we have a gazillion companies wanting to implement their guardrails version and we need to 'productize' it then we can discuss that point.

Agree, that is also the stance taken in the SSTable read/writer thread; we are not ready to support that but if there is enough demand we can justify the work.